### ISSUE TITLE
Topology visualization

### ISSUE THREAD

**Issue by carstenbauer** [JuliaParallel/Hwloc.jl#26]
Hey there,

it would be cool if one could get a nice visualization of the `topology_load()` information, as the latter is somewhat hard to parse (especially on cluster nodes).

As `hwloc` comes with `lstopo`, which already does this, we could perhapas temporarily  go with something along the lines of
```julia
lstopo = joinpath(dirname(Hwloc.libhwloc), "lstopo")
Sys.iswindows() && (lstopo = string(lstopo, ".exe"))
run(`$lstopo`)
```
(tested on my Windows 10 system)

Of course, it would be nicer to not have to call another binary for this and just parse and plot the topology directly in Julia (via Plots.jl for example). I might be able to post something later today, but can't promise anything.

Best
Carsten

**Comment by carstenbauer** [comment]
I just noticed that on my linux machine lstopo is not downloaded as part of the build process, so the suggestion above might only work for Windows?

**Comment by carstenbauer** [comment]
Simple solution:

```julia
using Hwloc, AbstractTrees, Plots, GraphRecipes
t = topology_load()

AbstractTrees.children(x::Hwloc.Object) = x.children
AbstractTrees.printnode(io::IO, x::Hwloc.Object) = x.type_ in (:Core, :PU) ? print(io, "$(x.type_) $(x.logical_index)") : print(io, x.type_)

default(size=(800, 800))
plot(TreePlot(t), fontsize=8, nodeshape=:circle, nodecolor=:white)
```

![topology](https://user-images.githubusercontent.com/187980/112768790-06a7a200-901e-11eb-894c-8fe71680e77f.png)

I also like a more simplistic version of the default printing:

```Julia
function AbstractTrees.printnode(io::IO, x::Hwloc.Object)
    if x.type_ in (:Core, :PU)
        print(io, "$(x.type_) $(x.logical_index)")
    elseif x.type_ in (:L3Cache, :L2Cache, :L1Cache)
        str = string(x.type_)[1:2]
        print(io, "$str")
    else
        print(io, x.type_)
    end
end

print_tree(t, maxdepth=10)
```

```
Machine
└─ Package
   └─ L3
      ├─ L2
      │  └─ L1
      │     └─ Core 0
      │        ├─ PU 0
      │        └─ PU 1
      ├─ L2
      │  └─ L1
      │     └─ Core 1
      │        ├─ PU 2
      │        └─ PU 3
      ├─ L2
      │  └─ L1
      │     └─ Core 2
      │        ├─ PU 4
      │        └─ PU 5
      ├─ L2
      │  └─ L1
      │     └─ Core 3
      │        ├─ PU 6
      │        └─ PU 7
      ├─ L2
      │  └─ L1
      │     └─ Core 4
      │        ├─ PU 8
      │        └─ PU 9
      └─ L2
         └─ L1
            └─ Core 5
               ├─ PU 10
               └─ PU 11
```


Here for a cluster node:

```
Machine
├─ Socket
│  └─ L3
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 0
│     │        └─ PU 0
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 1
│     │        └─ PU 1
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 2
│     │        └─ PU 2
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 3
│     │        └─ PU 3
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 4
│     │        └─ PU 4
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 5
│     │        └─ PU 5
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 6
│     │        └─ PU 6
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 7
│     │        └─ PU 7
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 8
│     │        └─ PU 8
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 9
│     │        └─ PU 9
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 10
│     │        └─ PU 10
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 11
│     │        └─ PU 11
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 12
│     │        └─ PU 12
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 13
│     │        └─ PU 13
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 14
│     │        └─ PU 14
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 15
│     │        └─ PU 15
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 16
│     │        └─ PU 16
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 17
│     │        └─ PU 17
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 18
│     │        └─ PU 18
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 19
│     │        └─ PU 19
│     ├─ L2
│     │  └─ L1
│     │     └─ Core 20
│     │        └─ PU 20
│     └─ L2
│        └─ L1
│           └─ Core 21
│              └─ PU 21
└─ Socket
   └─ L3
      ├─ L2
      │  └─ L1
      │     └─ Core 22
      │        └─ PU 22
      ├─ L2
      │  └─ L1
      │     └─ Core 23
      │        └─ PU 23
      ├─ L2
      │  └─ L1
      │     └─ Core 24
      │        └─ PU 24
      ├─ L2
      │  └─ L1
      │     └─ Core 25
      │        └─ PU 25
      ├─ L2
      │  └─ L1
      │     └─ Core 26
      │        └─ PU 26
      ├─ L2
      │  └─ L1
      │     └─ Core 27
      │        └─ PU 27
      ├─ L2
      │  └─ L1
      │     └─ Core 28
      │        └─ PU 28
      ├─ L2
      │  └─ L1
      │     └─ Core 29
      │        └─ PU 29
      ├─ L2
      │  └─ L1
      │     └─ Core 30
      │        └─ PU 30
      ├─ L2
      │  └─ L1
      │     └─ Core 31
      │        └─ PU 31
      ├─ L2
      │  └─ L1
      │     └─ Core 32
      │        └─ PU 32
      ├─ L2
      │  └─ L1
      │     └─ Core 33
      │        └─ PU 33
      ├─ L2
      │  └─ L1
      │     └─ Core 34
      │        └─ PU 34
      ├─ L2
      │  └─ L1
      │     └─ Core 35
      │        └─ PU 35
      ├─ L2
      │  └─ L1
      │     └─ Core 36
      │        └─ PU 36
      ├─ L2
      │  └─ L1
      │     └─ Core 37
      │        └─ PU 37
      ├─ L2
      │  └─ L1
      │     └─ Core 38
      │        └─ PU 38
      ├─ L2
      │  └─ L1
      │     └─ Core 39
      │        └─ PU 39
      ├─ L2
      │  └─ L1
      │     └─ Core 40
      │        └─ PU 40
      ├─ L2
      │  └─ L1
      │     └─ Core 41
      │        └─ PU 41
      ├─ L2
      │  └─ L1
      │     └─ Core 42
      │        └─ PU 42
      └─ L2
         └─ L1
            └─ Core 43
               └─ PU 43
```

**Comment by carstenbauer** [comment]
Of course, we can't add this because we'd have to introduce many dependencies (which IMO isn't worth it). However, maybe we could create a sub package `TopologyPlot` or `HwlocPlot` in this repository similar to what they are doing over at BenchmarkTools.jl with BenchmarkPlots, see https://github.com/JuliaCI/BenchmarkTools.jl/pull/194.

@eschnett Would you consider merging such a PR?

**Comment by eschnett** [comment]
I would be delighted to!

Of course, creating a new package in a new repository would also work – as you prefer. I'd be happy to add an example to the README that points to and uses this package, and shows the resulting figure.

**Comment by carstenbauer** [comment]
I will draft something in a new repo and we can relocate the code later if necessary.

**Comment by carstenbauer** [comment]
FYI: https://github.com/crstnbr/HwlocViz.jl

**Comment by carstenbauer** [comment]
Now that we have `topology_graphical()` I'm closing this. If someone wants to implement native printing of the topology please go for it in a new issue.
